File: IRemember/app/src/main/java/ru/goodibunakov/iremember/MainActivity.java
package ru.goodibunakov.iremember;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentManager;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.goodibunakov.iremember.adapter.TabAdapter;
import ru.goodibunakov.iremember.alarm.AlarmHelper;
import ru.goodibunakov.iremember.database.DbHelper;
import ru.goodibunakov.iremember.dialog.AddingTaskDialogFragment;
import ru.goodibunakov.iremember.fragment.CurrentTaskFragment;
import ru.goodibunakov.iremember.fragment.DoneTaskFragment;
import ru.goodibunakov.iremember.fragment.SplashFragment;
import ru.goodibunakov.iremember.fragment.TaskFragment;
import ru.goodibunakov.iremember.model.ModelTask;

public class MainActivity extends AppCompatActivity implements AddingTaskDialogFragment.AddingTaskListener,
        DoneTaskFragment.OnTaskRestoreListener, CurrentTaskFragment.OnTaskDoneListener {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.tab_layout)
    TabLayout tabLayout;
    @BindView(R.id.pager)
    ViewPager viewPager;
    @BindView(R.id.search_view)
    SearchView searchView;

    FragmentManager fragmentManager;
    PreferenceHelper preferenceHelper;
    TabAdapter tabAdapter;
    TaskFragment currentTaskFragment;
    TaskFragment doneTaskFragment;
    SplashFragment splashFragment;

    public DbHelper dbHelper;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        PreferenceHelper.getInstance().init(getApplicationContext());
        preferenceHelper = PreferenceHelper.getInstance();

        AlarmHelper.getInstance().init(getApplicationContext());

        dbHelper = new DbHelper(getApplicationContext());

        fragmentManager = getSupportFragmentManager();
        runSplash();
        setUI();
    }

    @Override
    protected void onResume() {
        super.onResume();
        RememberApp.activityResumed();
    }

    @Override
    protected void onPause() {
        super.onPause();
        RememberApp.activityPaused();
    }

    public void runSplash() {
        if (!preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE)) {
            splashFragment = new SplashFragment();
            fragmentManager.beginTransaction()
                    .replace(R.id.content_frame, splashFragment)
                    .addToBackStack(null)
                    .commit();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_splash) {
            item.setChecked(!item.isChecked());
            preferenceHelper.putBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE, item.isChecked());
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        MenuItem splashItem = menu.findItem(R.id.action_splash);
        splashItem.setChecked(preferenceHelper.getBoolean(PreferenceHelper.SPLASH_IS_INVISIBLE));
        return true;
    }

    private void setUI() {
        toolbar.setTitleTextColor(ContextCompat.getColor(MainActivity.this, android.R.color.white));
        toolbar.setTitleMarginStart(68);
        toolbar.setLogo(R.drawable.toolbar_icon);
        setSupportActionBar(toolbar);

        tabLayout.addTab(tabLayout.newTab().setText(R.string.current_task));
        tabLayout.addTab(tabLayout.newTab().setText(R.string.done_task));

        tabAdapter = new TabAdapter(fragmentManager, 2);

        viewPager.setAdapter(tabAdapter);
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));

        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                viewPager.setCurrentItem(tab.getPosition());
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });

        currentTaskFragment = (CurrentTaskFragment) tabAdapter.getItem(TabAdapter.CURRENT_TASK_FRAGMENT_POSITION);
        doneTaskFragment = (DoneTaskFragment) tabAdapter.getItem(TabAdapter.DONE_TASK_FRAGMENT_POSITION);

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                currentTaskFragment.findTasks(newText);
                doneTaskFragment.findTasks(newText);
                return false;
            }
        });
    }

    @Override
    public void onTaskAdded(ModelTask newTask) {
        Log.d("debug", "currentTaskFragment = " + currentTaskFragment);
        currentTaskFragment.addTask(newTask, true);
    }

    @Override
    public void onTaskAddingCancel() {
//        Toast.makeText(this, getResources().getString(R.string.task_canceled), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onTaskDone(ModelTask modelTask) {
        doneTaskFragment.addTask(modelTask, false);
    }

    @Override
    public void onTaskRestore(ModelTask modelTask) {
        currentTaskFragment.addTask(modelTask, false);
    }
}

File: IRemember/app/src/main/java/ru/goodibunakov/iremember/presentation/view/activity/MainActivity.kt
package ru.goodibunakov.iremember.presentation.view.activity

import android.os.Bundle
import android.os.PersistableBundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.widget.SearchView
import androidx.core.content.ContextCompat
import androidx.viewpager2.widget.ViewPager2
import by.kirich1409.viewbindingdelegate.viewBinding
import moxy.MvpAppCompatActivity
import moxy.ktx.moxyPresenter
import ru.goodibunakov.iremember.R
import ru.goodibunakov.iremember.RememberApp
import ru.goodibunakov.iremember.databinding.ActivityMainBinding
import ru.goodibunakov.iremember.presentation.bus.QueryEvent
import ru.goodibunakov.iremember.presentation.presenter.MainActivityPresenter
import ru.goodibunakov.iremember.presentation.view.adapter.TabAdapter2
import ru.goodibunakov.iremember.presentation.view.dialog.DeleteDoneTasksDialogFragment
import ru.goodibunakov.iremember.presentation.view.fragment.SplashFragment


class MainActivity : MvpAppCompatActivity(R.layout.activity_main), MainActivityView {

    private lateinit var onPageChangeCallback: ViewPager2.OnPageChangeCallback
    private val binding by viewBinding(ActivityMainBinding::bind)

    private val mainActivityPresenter by moxyPresenter {
        MainActivityPresenter(
            RememberApp.sharedPreferencesRepository,
            RememberApp.getEventBus(),
            RememberApp.databaseRepository
        )
    }

    companion object {
        const val TOOLBAR_TITLE_PADDING = 2
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d("debug", "MainActivity onCreate")
    }

    override fun onResume() {
        super.onResume()
        RememberApp.activityResumed()
    }

    override fun onPause() {
        super.onPause()
        RememberApp.activityPaused()
    }

    override fun runSplash() {
        supportFragmentManager.beginTransaction()
            .replace(R.id.content_frame, SplashFragment())
            .addToBackStack(null)
            .commit()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (val id = item.itemId) {
            R.id.action_splash -> {
                mainActivityPresenter.itemSelected(id)
                return true
            }

            R.id.action_trash -> {
                mainActivityPresenter.itemSelected(id)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun itemSelected(id: Int) {
        when (id) {
            R.id.action_splash -> {
                val splashItem = binding.toolbar.menu.findItem(id)
                splashItem.isChecked = !splashItem.isChecked
                mainActivityPresenter.saveBoolean(splashItem.isChecked)
            }

            R.id.action_trash -> {
                mainActivityPresenter.showDeleteDoneTasksDialog()
            }
        }
    }

    override fun showDeleteDoneTasksDialog() {
        val dialog = DeleteDoneTasksDialogFragment()
        dialog.show(supportFragmentManager, "DeleteDoneTasksDialogFragment")
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        Log.d("debug", "onCreateOptionsMenu")
        menuInflater.inflate(R.menu.menu_main, menu)
        val splashItem = menu.findItem(R.id.action_splash)
        mainActivityPresenter.setSplashItemChecked(splashItem.itemId)
        mainActivityPresenter.setDeleteAllDoneTasksIconVisible()
        return true
    }

    override fun setSplashItemState(id: Int, isChecked: Boolean) {
        binding.toolbar.menu?.findItem(id)?.isChecked = isChecked
    }

    override fun setUI() {
        Log.d("debug", "setUI")
        setToolbar()
        setBottomMenuAndViewpager()

        binding.searchView.setOnQueryTextListener(object : OnQueryTextListener {
            override fun onQueryTextChange(newText: String): Boolean {
                Log.d("rx", "onQueryTextChange $newText")
                mainActivityPresenter.find(QueryEvent(newText))
                return true
            }
        })
        binding.searchView.setOnCloseListener {
            Log.d("debug", "close")
            mainActivityPresenter.find(QueryEvent(""))
            false
        }
    }

    override fun onDestroy() {
        mainActivityPresenter.onDestroyView()
        super.onDestroy()
    }

    override fun onDestroyView() {
        binding.searchView.setOnQueryTextListener(null)
        binding.viewPager2.unregisterOnPageChangeCallback(onPageChangeCallback)
    }

    private fun setToolbar() {
        binding.toolbar.setTitleTextColor(ContextCompat.getColor(this, android.R.color.white))
        binding.toolbar.setLogo(R.drawable.toolbar_icon)
        binding.toolbar.titleMarginStart = ContextCompat.getDrawable(
            this,
            R.drawable.toolbar_icon
        )!!.intrinsicWidth + TOOLBAR_TITLE_PADDING
        setSupportActionBar(binding.toolbar)
    }

    private fun setBottomMenuAndViewpager() {
        binding.viewPager2.adapter = TabAdapter2(this)
        binding.bottomMenu.setupWithViewPager2(binding.viewPager2)
        onPageChangeCallback = object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                mainActivityPresenter.showDeleteAllTasksIcon(position == 1)
            }
        }
        binding.viewPager2.registerOnPageChangeCallback(onPageChangeCallback)
    }


    override fun showDeleteAllTasksIcon(visible: Boolean) {
        binding.toolbar.menu?.findItem(R.id.action_trash)?.isVisible = visible
    }

    override fun onSaveInstanceState(outState: Bundle, outPersistentState: PersistableBundle) {}

    interface OnQueryTextListener : SearchView.OnQueryTextListener {
        override fun onQueryTextSubmit(query: String): Boolean {
            return false
        }
    }
}